File: src/main/java/io/beekeeper/bots/pizza/OrderHelper.java
package io.beekeeper.bots.pizza;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;

import org.jdeferred2.Promise;
import org.jdeferred2.impl.DeferredObject;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;

import io.beekeeper.bots.pizza.shell.ProcessExecutor;

public class OrderHelper {

    public static Promise<Void, Void, Void> executeOrder(Collection<OrderItem> orderItems, boolean dryRun) {
        DeferredObject<Void, Void, Void> deferred = new DeferredObject<>();

        List<String> command = new ArrayList<>(Arrays.asList(
                "node",
                "pizza-ordering/app.js",
                toJSON(orderItems).toString()
        ));
        if (!dryRun) {
            command.add("-x");
        }

        new Thread(() -> {
            try {
                System.out.println("command = " + command);
                ProcessExecutor.CommandResult result = ProcessExecutor.executeCommand(command);
                if (result.getExitCode() == 0) {
                    deferred.resolve(null);
                } else {
                    deferred.reject(null);
                }
            } catch (IOException | InterruptedException e) {
                e.printStackTrace();
            }
        }).start();

        return deferred.promise();
    }

    private static JsonArray toJSON(Collection<OrderItem> orderItems) {
        JsonArray jsonArray = new JsonArray();
        orderItems.forEach(item -> jsonArray.add(toJSON(item)));
        return jsonArray;
    }

    private static JsonObject toJSON(OrderItem item) {
        String articleNumber = item.getMenuItem().getArticleNumber();
        if (item.getMenuItem().getParentArticleNumber() != null) {
            articleNumber = item.getMenuItem().getParentArticleNumber();
        }

        JsonObject json = new JsonObject();
        json.addProperty("articleId", item.getMenuItem().getArticleId());
        json.addProperty("articleNumber", articleNumber);
        json.addProperty("commodityGroupId", item.getMenuItem().getCommodityGroupId());
        return json;
    }
}

File: src/main/java/io/beekeeper/bots/pizza/OrderHelper.kt
package io.beekeeper.bots.pizza

import com.google.gson.JsonArray
import com.google.gson.JsonObject
import io.beekeeper.bots.pizza.extensions.logger
import io.beekeeper.bots.pizza.extensions.mapIf
import io.beekeeper.bots.pizza.extensions.runAsync
import io.beekeeper.bots.pizza.shell.ProcessExecutor
import org.jdeferred2.Promise
import org.jdeferred2.impl.DeferredObject

object OrderHelper {

    private val log = logger()

    fun executeOrder(orderItems: Collection<OrderItem>, dryRun: Boolean): Promise<Unit, Unit, Unit> {
        val deferred = DeferredObject<Unit, Unit, Unit>()
        val command = generateCommand(orderItems, dryRun)
        runAsync {
            try {
                log.info("command = $command")
                ProcessExecutor.executeCommand(command)
                        .let { result ->
                            if (result.exitCode == 0) {
                                deferred.resolve(Unit)
                            } else {
                                deferred.reject(Unit)
                            }
                        }
            } catch (e: Exception) {
                log.error("Failed to submit order", e)
                if (deferred.isPending) {
                    deferred.reject(Unit)
                }
            }
        }

        return deferred.promise()
    }

    private fun generateCommand(orderItems: Collection<OrderItem>, dryRun: Boolean) =
            listOf(
                    "node",
                    "pizza-ordering/app.js",
                    toJSON(orderItems).toString()
            ).mapIf(!dryRun) {
                it.plus("-x")
            }

    private fun toJSON(orderItems: Collection<OrderItem>) =
            JsonArray().apply {
                orderItems.forEach { item -> add(toJSON(item)) }
            }

    private fun toJSON(item: OrderItem) = JsonObject().apply {
        addProperty("articleId", item.menuItem.articleId)
        addProperty("articleNumber", item.menuItem.parentArticleNumber ?: item.menuItem.articleNumber)
        addProperty("commodityGroupId", item.menuItem.commodityGroupId)
    }
}
